src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlyFollowedBySemicolonHasLineBreakAfter.java

```diff
@@ -0,0 +1,76 @@
+/*
+RightCurly
+option = ALONE_OR_SINGLELINE
+tokens = LITERAL_TRY , LITERAL_CATCH , LITERAL_FINALLY , LITERAL_IF , LITERAL_ELSE , CLASS_DEF , \
+         METHOD_DEF , CTOR_DEF , LITERAL_FOR , LITERAL_WHILE , LITERAL_DO , STATIC_INIT , \
+         INSTANCE_INIT , ANNOTATION_DEF , ENUM_DEF , INTERFACE_DEF , RECORD_DEF , COMPACT_CTOR_DEF \
+         , OBJBLOCK
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.blocks.rightcurly;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+
+import java.util.*;
+
+public class InputRightCurlyFollowedBySemicolonHasLineBreakAfter {
+        Object obj_anon = new Object() {
+        }; int i = 0; // violation '';' at column 10 should have line break after'
+
+        void method1() {
+        } int j = 10; // violation ''}' at column 9 should be alone on a line'
+
+        class SomeClass {
+        } int k = 4; // violation ''}' at column 9 should be alone on a line'
+
+        Object obj_anon2 = new Object() {
+        void anonInnerMethod(){
+                System.out.println("Hello world");
+        }
+        }; int i2 = 0; // violation '';' at column 10 should have line break after'
+
+        Runnable r = new Runnable() {
+        @Override
+        public void run() {
+        System.out.println("Hello from anonymous inner class!");
+        }
+        }; void methodOnSameLine() { // violation '';' at column 10 should have line break after'
+                // code here
+        }
+
+        Thread anotherT = new Thread() {
+        @Override
+        public void run() {
+                System.out.println("Hello from anonymous inner class!");
+        }
+        }; Thread t = new Thread() { // violation '';' at column 10 should have line break after'
+        @Override
+        public void run() {
+                System.out.println("Hello from anonymous inner class!");
+        }
+        }; int value = 99; // violation '';' at column 10 should have line break after'
+
+        public enum SomeEnum {
+
+        ENUM1 () {
+
+        },
+
+        ENUM2 () {
+
+        },
+        }
+
+        private static final Comparator<DetailAST> AST_LINE_COMPARATOR = new Comparator<>() {
+        @Override
+        public int compare(DetailAST aObj1, DetailAST aObj2) {
+            return aObj1.getLineNo() - aObj2.getLineNo();
+        }
+    };
```

Comment:
this is where the CI is pointing to ([https://checkstyle.semaphoreci.com/workflows/b4875b1b-7e59-4c75-8637-a721d6025ac1?pipeline_id=2e8f3f69-5e76-470e-80bf-82f9f9905ecd](https://checkstyle.semaphoreci.com/jobs/c43fa60a-2b80-4561-9cce-a8b59cbe11c5#L656:~:text=%5BERROR%5D%20%5Bcheckstyle%5D%20%5BERROR%5D%20/home/semaphore/checkstyle/.ci%2Dtemp/sevntu.checkstyle/sevntu%2Dchecks/src/main/java/com/github/sevntu/checkstyle/checks/coding/CustomDeclarationOrderCheck.java%3A255%3A5%3A%20%27%7D%27%20at%20column%205%20should%20be%20alone%20on%20a%20line.%20%5BRightCurly%5D)):

If I go to the sevntu repo, the following is where the violation is occuring (line 255):
https://github.com/shubh220922/sevntu.checkstyle/blob/0d01f49a18ba7507386517a51c9ebcd45d02f46d/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/CustomDeclarationOrderCheck.java#L247-L255
(image for the same)
![image](https://github.com/checkstyle/checkstyle/assets/90137881/4f0b32c9-2f45-45b7-8151-016aad089190)

Now what I did was copy the snippet of code and paste it in my inputfile that I was using for checking appropriate behaivour of my check. **It does not show violation there!**

Reply:
Is the config the same?

Reply:
@nrmancuso 
could you please elaborate your comment, i didn't understand.

I am using the following config for the input file that I have written:
```java
/*
RightCurly
option = ALONE_OR_SINGLELINE
tokens = LITERAL_TRY , LITERAL_CATCH , LITERAL_FINALLY , LITERAL_IF , LITERAL_ELSE , CLASS_DEF , \
         METHOD_DEF , CTOR_DEF , LITERAL_FOR , LITERAL_WHILE , LITERAL_DO , STATIC_INIT , \
         INSTANCE_INIT , ANNOTATION_DEF , ENUM_DEF , INTERFACE_DEF , RECORD_DEF , COMPACT_CTOR_DEF \
         , OBJBLOCK


*/
```

You can see I have included ``OBJBLOCK`` at the last.

Reply:
@nrmancuso please review.

Reply:
I am asking if the checkstyle config for this module that we run on sevntu is same configuration that you are using in your test.

Reply:
@nrmancuso thanks for the reply. Sevntu error fixed. 

This was the original config that I plainly edited to put ``OBJBLOCK`` property there. 
```xml
  <module name="RightCurly">
      <property name="tokens" value="METHOD_DEF"/>
      <property name="tokens" value="CTOR_DEF"/>
      <property name="tokens" value="CLASS_DEF"/>
      <property name="tokens" value="INSTANCE_INIT"/>
      <property name="tokens" value="LITERAL_FOR"/>
      <property name="tokens" value="STATIC_INIT"/>
      <property name="tokens" value="LITERAL_WHILE"/>
      <property name="tokens" value="LITERAL_CATCH"/>
      <property name="tokens" value="LITERAL_ELSE"/>
      <property name="tokens" value="LITERAL_FINALLY"/>
      <property name="tokens" value="LITERAL_IF"/>
      <property name="tokens" value="LITERAL_TRY"/>
      <property name="tokens" value="LITERAL_SWITCH"/>
      <property name="tokens" value="ANNOTATION_DEF"/>
      <property name="tokens" value="ENUM_DEF"/>
      <property name="tokens" value="RECORD_DEF"/>
      <property name="tokens" value="COMPACT_CTOR_DEF"/>
      <property name="tokens" value="OBJBLOCK"/>            <----- only this was present
      <property name="option" value="alone"/>
    </module>
    <module name="RightCurly">
      <property name="tokens" value="LITERAL_DO"/>
      <property name="option" value="same"/>
    </module>
    <module name="RightCurly">
      <property name="tokens" value="INTERFACE_DEF"/>
      <property name="option" value="alone_or_singleline"/>
    </module>
    <module name="RightCurly">
      <property name="tokens" value="INTERFACE_DEF"/>
      <property name="option" value="alone_or_singleline"/>
    </module>
```

I have modified it to:
```xml
<module name="RightCurly">
      <property name="tokens" value="METHOD_DEF"/>
      <property name="tokens" value="CTOR_DEF"/>
      <property name="tokens" value="CLASS_DEF"/>
      <property name="tokens" value="INSTANCE_INIT"/>
      <property name="tokens" value="LITERAL_FOR"/>
      <property name="tokens" value="STATIC_INIT"/>
      <property name="tokens" value="LITERAL_WHILE"/>
      <property name="tokens" value="LITERAL_CATCH"/>
      <property name="tokens" value="LITERAL_ELSE"/>
      <property name="tokens" value="LITERAL_FINALLY"/>
      <property name="tokens" value="LITERAL_IF"/>
      <property name="tokens" value="LITERAL_TRY"/>
      <property name="tokens" value="LITERAL_SWITCH"/>
      <property name="tokens" value="ANNOTATION_DEF"/>
      <property name="tokens" value="ENUM_DEF"/>
      <property name="tokens" value="RECORD_DEF"/>
      <property name="tokens" value="COMPACT_CTOR_DEF"/>
      <property name="option" value="alone"/>
    <module name="RightCurly">
      <property name="tokens" value="OBJBLOCK"/>
      <property name="option" value="alone_or_singleline"/>  <----- changed the config to add 
    </module>                                                       alone_or_singleline
    </module>
    <module name="RightCurly">
      <property name="tokens" value="LITERAL_DO"/>
      <property name="option" value="same"/>
    </module>
    <module name="RightCurly">
      <property name="tokens" value="INTERFACE_DEF"/>
      <property name="option" value="alone_or_singleline"/>
    </module>
```

Reply:
@nrmancuso @romani 

Fixing CI:
- [ ] coverage issue
- [ ] checkerFrameWork errors
- [ ] surviving mutation

Reply:
@nrmancuso need help here

### Problem with surviving mutants

#### Mutation 1

Mutant groovy report: 
```xml
Source File: "RightCurlyCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.blocks.RightCurlyCheck"
Method: "isRightcurlyFollowedBySemicolonHasLineBreakAfter"
Line Contents: "return tokenAfterTheNextToken != null"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 635
```

[Link to mutation report in CI](https://github.com/checkstyle/checkstyle/actions/runs/4988459602/jobs/8931211849?pr=12685#step:6:446)

This is happening in the method that I made: 
```java
    private static boolean isRightcurlyFollowedBySemicolonHasLineBreakAfter(RightCurlyOption
                                                                              bracePolicy,
                                                                              Details details) {
        final DetailAST tokenAfterTheNextToken = Details.getNextToken(details.nextToken);

        return tokenAfterTheNextToken != null
            && bracePolicy == RightCurlyOption.ALONE_OR_SINGLELINE
            && isRightcurlyFollowedBySemicolon(details)
            && TokenUtil.areOnSameLine(details.rcurly, tokenAfterTheNextToken);
    }
```

Now I am confused as to what mutation is PIT making at all? 
**_removed conditional - replaced equality check with true_** : when I read this I assume some code to be like ``a != b`` changed to ``true`` and in the above case ``return tokenAfterTheNextToken != null  && ......`` changed to ``return true && .....`` but, when I check locally and hardcode the mutation, some of my test cases DO fail indicating the the mutation is killed but PIT report says otherwise. I am confused here. 

#### Mutation 2

Mutant groovy report:
```xml
Source File: "RightCurlyCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.blocks.RightCurlyCheck$Details"
Method: "getDetailsForAnonymousInnerClass"
Line Contents: "lcurly = ast.getFirstChild();"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.experimental.NakedReceiverMutator"
Description: "replaced call to com/puppycrawl/tools/checkstyle/api/DetailAST::getFirstChild with receiver"
Line Number: 860
``` 
[Link to mutant report in CI.](https://github.com/checkstyle/checkstyle/actions/runs/4988459602/jobs/8931211849?pr=12685#step:6:454)

Java code where the mutation is occuring:
```java 
private static Details getDetailsForAnonymousInnerClass(DetailAST ast) {

            final DetailAST rcurly;
            final DetailAST lcurly;
            final DetailAST nextToken;
            final boolean checkStatusOfAncestor = checkIfAncestorIsVariableDef(ast);

            if (checkStatusOfAncestor) {
                rcurly = ast.getLastChild();
                lcurly = ast.getFirstChild();
                nextToken = getNextToken(ast);
            }
            else {
                rcurly = null;
                lcurly = null;
                nextToken = null;
            }

            return new Details(lcurly, rcurly, nextToken, false);
        }
```
Now here is the thing, if we closely look at the AST of ``OBJBLOCK`` from different code snippets:
```
                        |--RPAREN -> ) [16:37]
        |   |           `--OBJBLOCK -> OBJBLOCK [16:39]
        |   |               |--LCURLY -> { [16:39]
        |   |               `--RCURLY -> } [17:8]
        |   `--SEMI -> ; [17:9]
```
It has a similar structure to the above AST. What i want to say is if we closely look at the ``LCURLY`` position, we see that the ``OBJBLOCK`` starts at exactly the same place where the ``LCURLY`` start. This is a problem as the mutation shown above is surviving because of this. 

PIT is changing the line ``lcurly = ast.getFirstChild();`` with receiver that is, ``lcurly = ast`` and, according to how I've designed the method, ``ast`` contains ``OBJBLOCK`` and the first child of ``ast`` is ``LCURLY``. Since, position of ``LCURLY`` and ``OBJBLOCK`` is always the same, the mutation is surviving and I don't think this mutation will be killed through input file at all. 

Only way to kill this mutation is to change the code and include a small check somewhere in the code which does something like ``TokenTypes.LCURLY == details.lcurly.getType()`` to cross check whether the lcurly is indeed a lcurly and not ``OBJBLOCK`` (as changing it with receiver sends ``OBJBLOCK`` instead of ``LCURLY`` in the code).

I infact did the above and killed the mutation but, this resulted in additional few mutation arising. If I at all write ``TokenTypes.LCURLY == details.lcurly.getType()``, this line itself will cause a mutation to survive because ``lcurly``  has ``LCURLY`` all the time.

How can I fix this. @nrmancuso

Reply:
For 1: I would check to make due that pitest is not actually mutating the second condition.
For 2: If we know that lcurly is always the first child of objblock, why bother with any checks at all?

Reply:
@nrmancuso DONE.

### 1.

Thanks. The second line was where the mutation was surviving. Killed it. 

### 2. 

For complete clarity, we need to look at every possible case where an ``OBJBLOCK`` occurs and note the corresponding position of ``LCURLY``. The below image contains the exhaustive set of all places where ``OBJBLOCK`` occurs:

![image](https://github.com/checkstyle/checkstyle/assets/90137881/4e5be4a5-1f6c-43c8-9479-a0563121cfb2)
(Image is of ``OBJBLOCK`` javadoc from ``TokenTypes.java``)

1. Class
2. Interface
3. Enum
4. Annotation
5. Enum Constant Declaration
6. Anonymous Inner Declarations

The respective ASTs of the above mentioned list are below:

### 1. Class
```
|--CLASS_DEF -> CLASS_DEF [22:8]
        |   |--MODIFIERS -> MODIFIERS [22:8]
        |   |--LITERAL_CLASS -> class [22:8]
        |   |--IDENT -> SomeClass [22:14]
        |   `--OBJBLOCK -> OBJBLOCK [22:24]
        |       |--LCURLY -> { [22:24]
        |       `--RCURLY -> } [23:8]
```
### 2. Interface
```
|--INTERFACE_DEF -> INTERFACE_DEF [70:8]
        |   |--MODIFIERS -> MODIFIERS [70:8]
        |   |   `--LITERAL_PUBLIC -> public [70:8]
        |   |--LITERAL_INTERFACE -> interface [70:15]
        |   |--IDENT -> Drawable [70:25]
        |   `--OBJBLOCK -> OBJBLOCK [70:34]
        |       |--LCURLY -> { [70:34]
        |       |--METHOD_DEF -> METHOD_DEF [71:16]
        |       |   |--MODIFIERS -> MODIFIERS [71:16]
        |       |   |--TYPE -> TYPE [71:16]
        |       |   |   `--LITERAL_VOID -> void [71:16]
        |       |   |--IDENT -> draw [71:21]
        |       |   |--LPAREN -> ( [71:25]
        |       |   |--PARAMETERS -> PARAMETERS [71:26]
        |       |   |--RPAREN -> ) [71:26]
        |       |   `--SEMI -> ; [71:27]
        |       `--RCURLY -> } [72:8]
        `--RCURLY -> } [75:0]
```
### 3. Enum
```
 |--ENUM_DEF -> ENUM_DEF [52:8]
        |   |--MODIFIERS -> MODIFIERS [52:8]
        |   |   `--LITERAL_PUBLIC -> public [52:8]
        |   |--ENUM -> enum [52:15]
        |   |--IDENT -> SomeEnum [52:20]
        |   `--OBJBLOCK -> OBJBLOCK [52:29]
        |       |--LCURLY -> { [52:29]
        |       |--ENUM_CONSTANT_DEF -> ENUM_CONSTANT_DEF [54:8]
        |       |   |--ANNOTATIONS -> ANNOTATIONS [54:8]
        |       |   |--IDENT -> ENUM1 [54:8]
        |       |   |--LPAREN -> ( [54:14]
        |       |   |--ELIST -> ELIST [54:15]
        |       |   |--RPAREN -> ) [54:15]
        |       |   `--OBJBLOCK -> OBJBLOCK [54:17]
        |       |       |--LCURLY -> { [54:17]
        |       |       `--RCURLY -> } [56:8]
        |       |--COMMA -> , [56:9]
        |       |--ENUM_CONSTANT_DEF -> ENUM_CONSTANT_DEF [58:8]
        |       |   |--ANNOTATIONS -> ANNOTATIONS [58:8]
        |       |   |--IDENT -> ENUM2 [58:8]
        |       |   |--LPAREN -> ( [58:14]
        |       |   |--ELIST -> ELIST [58:15]
        |       |   |--RPAREN -> ) [58:15]
        |       |   `--OBJBLOCK -> OBJBLOCK [58:17]
        |       |       |--LCURLY -> { [58:17]
        |       |       `--RCURLY -> } [60:8]
        |       |--COMMA -> , [60:9]
        |       `--RCURLY -> } [61:8]
```
### 4. Annotation
```
|--ANNOTATION_DEF -> ANNOTATION_DEF [70:8]
        |   |--MODIFIERS -> MODIFIERS [70:8]
        |   |   `--LITERAL_PUBLIC -> public [70:8]
        |   |--AT -> @ [70:15]
        |   |--LITERAL_INTERFACE -> interface [70:16]
        |   |--IDENT -> MyAnnotation [70:26]
        |   `--OBJBLOCK -> OBJBLOCK [71:8]
        |       |--LCURLY -> { [71:8]
        |       |--ANNOTATION_FIELD_DEF -> ANNOTATION_FIELD_DEF [72:12]
        |       |   |--MODIFIERS -> MODIFIERS [72:12]
        |       |   |--TYPE -> TYPE [72:12]
        |       |   |   `--LITERAL_INT -> int [72:12]
        |       |   |--IDENT -> someValue [72:16]
        |       |   |--LPAREN -> ( [72:25]
        |       |   |--RPAREN -> ) [72:26]
        |       |   `--SEMI -> ; [72:27]
        |       `--RCURLY -> } [73:8]
        `--RCURLY -> } [75:0]
```

### 5. Enum Constant Declaration **_Same as point 3_**

### 6. Anonymous Inner Declarations _**This is the case we are dealing with in this PR**_
```
  |--VARIABLE_DEF -> VARIABLE_DEF [16:8]
        |   |--MODIFIERS -> MODIFIERS [16:8]
        |   |--TYPE -> TYPE [16:8]
        |   |   `--IDENT -> Object [16:8]
        |   |--IDENT -> obj_anon [16:15]
        |   |--ASSIGN -> = [16:24]
        |   |   `--EXPR -> EXPR [16:26]
        |   |       `--LITERAL_NEW -> new [16:26]
        |   |           |--IDENT -> Object [16:30]
        |   |           |--LPAREN -> ( [16:36]
        |   |           |--ELIST -> ELIST [16:37]
        |   |           |--RPAREN -> ) [16:37]
        |   |           `--OBJBLOCK -> OBJBLOCK [16:39]
        |   |               |--LCURLY -> { [16:39]
        |   |               `--RCURLY -> } [17:8]
        |   `--SEMI -> ; [17:9]
```
### Conclusion for Mutant 2 

**It is safe to let this mutation exist for now as ``LCURLY`` is _**ALWAYS**_  the first child of ``OBJBLOCK``.**